### PR TITLE
Add live tail methods

### PR DESCRIPTION
**Pull request by s2i-sfx**

### CODE REVIEW

File: lib/client/signalflow/request_manager.js [hunk 1]
```diff
@@ -390,11 +390,17 @@ function RequestManager(options) {
     //destroy all open SSE connections
   }
 
+  function registerLiveTail(opts, callbacks, overrideRequestId ){
+    var requestId = overrideRequestId || getRequestIdAndIncrement();
+    streamComputationWebsocket(opts, requestId, callbacks);
```
Comment (s2i-sfx) [inline lib/client/signalflow/request_manager.js:395]: @mpetazzoni I added this new method here to register a live tail as we spoke. 

I am confused by what is expected by the opts input argument. Looks like addComputation method which is called by streamComputationWebsocket needs a params.range as per https://github.com/signalfx/signalfx-nodejs/blob/master/lib/client/signalflow/request_manager.js#L233

Also, looks like a requestType and isRetryPatchMode are needed here. What do these arguments mean and what do I provide for these?
 https://github.com/signalfx/signalfx-nodejs/blob/master/lib/client/signalflow/request_manager.js#L259

Comment (mpetazzoni) [reply lib/client/signalflow/request_manager.js:395]: You shouldn't use `streamComputationWebsocket()`, this is specifically for starting a Signalflow computation. You'll want to send your own request into the Websocket (with `activeSocket.send(...)`, and make sure you register something into the `knownComputations` map.

File: lib/client/signalflow/signalflow_client.js [hunk 2]
```diff
@@ -1,8 +1,7 @@
-'use strict';
+"use strict";
```
Comment (mpetazzoni) [inline lib/client/signalflow/signalflow_client.js:1]: Could you keep all those stylistic changes to another commit in this PR?

File: lib/client/signalflow/signalflow_client.js [hunk 3]
```diff
@@ -63,28 +62,78 @@ function SignalflowClient(apiToken, options) {
...
     explain: function (opts) {
-      return signalflowRequest(opts, 'explain');
+      return signalflowRequest(opts, "explain");
     },
     preflight: function (opts) {
-      return signalflowRequest(opts, 'preflight');
-    }
+      return signalflowRequest(opts, "preflight");
+    },
+    liveTail: function (opts) {
```
Comment (mpetazzoni) [inline lib/client/signalflow/signalflow_client.js:134]: You can change that to simply:

```js
  liveTail: liveTailRequest,
```

File: lib/client/signalflow/signalflow_client.js [hunk 4]
```diff
@@ -63,28 +62,78 @@ function SignalflowClient(apiToken, options) {
...
     };
   }
 
+  function liveTailRequest(opts) {
+    var msgBuffer = [];
+    var callback = null;
+
+    function resolveMessage(msg) {
+      // we're duck typing error messages here, but it should be okay.
+      if (msg.hasOwnProperty("error") && !msg.type) {
```
Comment (mpetazzoni) [inline lib/client/signalflow/signalflow_client.js:82]: I don't think your message definition ever defines that field / error messages, does it?

File: lib/client/signalflow/request_manager.js (hunk 1 above)
Comment (s2i-sfx) [reply lib/client/signalflow/request_manager.js:395]: Ahh got it. Let me update

File: lib/client/signalflow/signalflow_client.js (hunk 3 above)
Comment (s2i-sfx) [reply lib/client/signalflow/signalflow_client.js:134]: ok I did that, but looks like I need to send in some opts? Because addComputation needs to require these 

`  var requiredParams = {
    execute: ['program'],
    explain: ['incidentId']
  };`

Comment (s2i-sfx) [reply lib/client/signalflow/signalflow_client.js:134]: where do these come from?

File: lib/client/signalflow/request_manager.js [hunk 5]
```diff
@@ -390,11 +390,22 @@ function RequestManager(options) {
     //destroy all open SSE connections
   }
 
+  function registerLiveTail(opts, callbacks, overrideRequestId ){
+    var requestId = overrideRequestId || getRequestIdAndIncrement();
+    addComputation(opts, requestId, transports.WEBSOCKET, callbacks, requestType, isRetryPatchMode);
```
Comment (s2i-sfx) [inline lib/client/signalflow/request_manager.js:395]: @mpetazzoni changed it to this, but can requestType and isRetryPatchMode be undefined?

Comment (s2i-sfx) [reply lib/client/signalflow/request_manager.js:395]: var requiredParams = { execute: ['program'], explain: ['incidentId'] };

File: lib/client/signalflow/signalflow_client.js (hunk 2 above)
Comment (s2i-sfx) [reply lib/client/signalflow/signalflow_client.js:1]: I think I auto added some of these unintentionally, let me revert.

File: lib/client/signalflow/request_manager.js (hunk 5 above)
Comment (s2i-sfx) [reply lib/client/signalflow/request_manager.js:395]: seems like these ^ are required

File: lib/client/signalflow/signalflow_client.js [hunk 6]
```diff
@@ -74,6 +74,53 @@ function SignalflowClient(apiToken, options) {
     };
   }
 
+  function liveTailRequest(opts) {
+    var msgBuffer = [];
+    var callback = null;
+
+    function resolveMessage(msg) {
+      // we're duck typing error messages here, but it should be okay.
+      if (msg.hasOwnProperty('error') && !msg.type) {
```
Comment (s2i-sfx) [inline lib/client/signalflow/signalflow_client.js:83]: I don;t think we need this, Im going to remove it.

File: lib/client/signalflow/signalflow_client.js (hunk 3 above)
Comment (mpetazzoni) [reply lib/client/signalflow/signalflow_client.js:126]: Those are for signalflow computation; you don't want to go through this path. Don't use `addComputation()` directly.

File: lib/client/signalflow/request_manager.js [hunk 7]
```diff
@@ -390,11 +390,34 @@ function RequestManager(options) {
     //destroy all open SSE connections
   }
 
+  function registerLiveTail(callbacks, overrideRequestId ){
+    var requestId = overrideRequestId || getRequestIdAndIncrement();
+    knownComputations[requestId] = {
+      callback: callbacks,
+      onMessage: router.onMessage,
```
Comment (mpetazzoni) [inline lib/client/signalflow/request_manager.js:397]: `router` isn't defined here. I don't think you need a `routedMessageHandler()` here anyway. Just set:

```js
onMessage: callbacks.onMessage,
```

You also don't need to set `callback` here, it won't be used.

File: lib/client/signalflow/request_manager.js [hunk 8]
```diff
@@ -390,11 +390,34 @@ function RequestManager(options) {
     //destroy all open SSE connections
   }
 
+  function registerLiveTail(callbacks, overrideRequestId ){
```
Comment (mpetazzoni) [inline lib/client/signalflow/request_manager.js:393]: I would expect you to need a `params` argument here to carry your live tail query.

File: lib/client/signalflow/request_manager.js [hunk 9]
```diff
@@ -390,11 +390,34 @@ function RequestManager(options) {
...
+    var requestId = overrideRequestId || getRequestIdAndIncrement();
+    knownComputations[requestId] = {
+      callback: callbacks,
+      onMessage: router.onMessage,
+      transport: transports.WEBSOCKET,
+      requestId: requestId,
+      lastDataBatchTime: router.getLatestBatchTimeStamp,
+    };
+    var jobObject = {};
```
Comment (mpetazzoni) [inline lib/client/signalflow/request_manager.js:402]: I'd avoid calling this `jobObject` to remove the confusion with signalflow jobs. Honestly just do:

```js
activeSocket.send(JSON.strigify({
  type: 'registerlivetail',
  channel: requestId,
  query: params.query,
  throttleOptions: params.throttleOptions,  // Assuming you'll be passing this too?
});

File: lib/client/signalflow/request_manager.js (hunk 8 above)
Comment (mpetazzoni) [inline lib/client/signalflow/request_manager.js:393]: Fix spacing around closing paren

File: lib/client/signalflow/request_manager.js [hunk 10]
```diff
@@ -390,11 +390,34 @@ function RequestManager(options) {
...
+    jobObject.channel = requestId;
+    activeSocket.send(JSON.stringify(jobObject));
+  }
+
+  function unRegisterLiveTail(requestId){
+    activeSocket.close();
```
Comment (mpetazzoni) [inline lib/client/signalflow/request_manager.js:408]: Don't close the socket! There may be other live tails or signaflow jobs using it. The RequestManager does not own the lifecycle of the WebSocket (the `SignalflowClient` does).

What you want to do instead is sent a `stoplivetail` request through the WebSocket.

File: lib/client/signalflow/request_manager.js [hunk 11]
```diff
@@ -390,11 +390,34 @@ function RequestManager(options) {
...
+  }
+
   return {
     authenticate: authenticate,
     execute: execute,
     stop: stop,
-    disconnect: disconnect
+    disconnect: disconnect,
+    registerLiveTail: registerLiveTail,
+    unRegisterLiveTail: unRegisterLiveTail
```
Comment (mpetazzoni) [inline lib/client/signalflow/request_manager.js:420]: Since the protocol calls it `stoplivetail` I would make the function name match that.

File: lib/client/signalflow/signalflow_client.js [hunk 12]
```diff
@@ -74,6 +74,44 @@ function SignalflowClient(apiToken, options) {
     };
   }
 
+  function liveTailRequest() {
```
Comment (mpetazzoni) [inline lib/client/signalflow/signalflow_client.js:77]: This function probably would be taking an `opts` argument that would contain the user's query.

File: lib/client/signalflow/signalflow_client.js [hunk 13]
```diff
@@ -74,6 +74,44 @@ function SignalflowClient(apiToken, options) {
...
+        callback(null, msg);
+      }
+    }
+
+    var callbacks = {
+      onMessage: msgCallback,
+      onError: errorMessageCallback,
+    };
+
+    var requestId = rm.registerLiveTail(callbacks, null);
```
Comment (mpetazzoni) [inline lib/client/signalflow/signalflow_client.js:98]: ... you would then pass these options to `rm.registerLiveTail()`

File: lib/client/signalflow/request_manager.js [hunk 14]
```diff
@@ -390,11 +390,47 @@ function RequestManager(options) {
...
 
+
+  function registerLiveTail(params, callbacks, overrideRequestId){
+    var requestId = overrideRequestId || getRequestIdAndIncrement();
+    knownComputations[requestId] = {
+      onMessage: callbacks.onMessage,
+      transport: transports.WEBSOCKET,
+      requestId: requestId,
+    };
+    knownComputations[requestId].streamController = getLiveTailStreamController();
```
Comment (mpetazzoni) [inline lib/client/signalflow/request_manager.js:401]: This won't work because your stream controller returned by `getLiveTailStreamController()` does not capture the arguments of the original request. If you look at https://github.com/signalfx/signalfx-nodejs/blob/master/lib/client/signalflow/request_manager.js#L75 and https://github.com/signalfx/signalfx-nodejs/blob/master/lib/client/signalflow/request_manager.js#L364 you'll see that `retry()` and `stop()` are called without arguments.

You need to capture them here with a closure IMHO:

```js
knownComputations[requestId] = {
  onMessage: callbacks.onMessage,
  transport: transports.WEBSOCKET,
  requestId: requestId,
  streamController: {
    retry: () => registerLiveTail(params, callbacks, overrideRequestId),
    stop: () => stopLiveTail(requestId),
  },
};
```

File: lib/client/signalflow/request_manager.js [hunk 15]
```diff
@@ -390,11 +390,47 @@ function RequestManager(options) {
...
+    activeSocket.send(JSON.stringify({
+      type: 'registerlivetail',
+      channel: requestId,
+      query: params.query,
+      throttleOptions: params.throttleOptions,
+    }));
+  }
+
+  function stopLiveTail(requestId){
+    activeSocket.send(JSON.stringify({
```
Comment (mpetazzoni) [inline lib/client/signalflow/request_manager.js:412]: You should check if `requestId` maps to an known computation; you can reuse `removeComputation()` to help:

```js
function stopLiveTail(requestId) {
  if (removeComputation(requestId)) {
    activeSocket.send(JSON.stringify({ ... });
  }
}
```

File: lib/client/signalflow/signalflow_client.js [hunk 16]
```diff
@@ -84,7 +130,8 @@ function SignalflowClient(apiToken, options) {
     },
     preflight: function (opts) {
       return signalflowRequest(opts, 'preflight');
-    }
+    },
+    liveTail: liveTailRequest,
```
Comment (mpetazzoni) [inline lib/client/signalflow/signalflow_client.js:134]: Nit: all lowercase `livetail` to match the casing of other request types.

File: lib/client/signalflow/request_manager.js (hunk 14 above)
Comment (s2i-sfx) [reply lib/client/signalflow/request_manager.js:401]: got it. yep. In my case I actually have arguments and need to close over them!!

File: lib/client/signalflow/request_manager.js (hunk 15 above)
Comment (s2i-sfx) [reply lib/client/signalflow/request_manager.js:412]: done

File: lib/client/signalflow/signalflow_client.js (hunk 16 above)
Comment (s2i-sfx) [reply lib/client/signalflow/signalflow_client.js:134]: done

File: lib/client/signalflow/request_manager.js (hunk 14 above)
Comment (s2i-sfx) [reply lib/client/signalflow/request_manager.js:401]: fixed

File: lib/client/signalflow/signalflow_client.js [hunk 17]
```diff
@@ -74,6 +74,52 @@ function SignalflowClient(apiToken, options) {
...
+      onError: errorMessageCallback
+    };
+
+    var requestId = rm.registerLiveTail(opts, callbacks, null);
+
+    return {
+      close: function () {
+        return rm.stopLiveTail(requestId);
+      },
+      stream: function (fn, errorFn) {
```
Comment (mpetazzoni) [inline lib/client/signalflow/signalflow_client.js:112]: I think `errorFn` doesn't end up being used (this is also a mistake in the `execute` path), and instead we do everything through the single callback function `fn` taking two arguments: `fn(error, msg)`.

If we are following the same pattern here, then we don't need this `errorFn` parameter.

File: lib/client/signalflow/signalflow_client.js [hunk 18]
```diff
@@ -74,6 +74,52 @@ function SignalflowClient(apiToken, options) {
...
+      if (!callback) {
+        msgBuffer.push(msg);
+      } else {
+        resolveMessage(msg);
+      }
+    }
+
+    var callbacks = {
+      onMessage: msgCallback,
+      onError: errorMessageCallback
```
Comment (mpetazzoni) [inline lib/client/signalflow/signalflow_client.js:103]: The request manager doesn't actually rely on your `onError` callback, so you don't need to pass it in. See my other comment about the method definition of `registerLiveTail()` anyway.

File: lib/client/signalflow/signalflow_client.js [hunk 19]
```diff
@@ -74,6 +74,52 @@ function SignalflowClient(apiToken, options) {
     };
   }
 
+  function liveTailRequest(opts) {
```
Comment (mpetazzoni) [inline lib/client/signalflow/signalflow_client.js:77]: Since this is part of the "public API" of the SignalflowClient, please add documentation for this method. In particular, describe what is expected to be in `opts`.

File: test/signalflow/request_manager.js [hunk 20]
```diff
@@ -7,13 +7,15 @@ var expect = require('chai').expect;
 describe('The SignalFlow Request Manager', function () {
   //TODO : re-enable this test once we figure out why it causes a hang
   var client = sflowclient('AUTHTOKEN');
-  it('should initialize a request manager with three methods, execute, stop, authenticate', function (done) {
+  it('should initialize a request manager with four methods, execute, stop, authenticate, registerLiveTail and stopLiveTail', function (done) {
```
Comment (mpetazzoni) [inline test/signalflow/request_manager.js:10]: So... is it four methods like the text says, five methods like the list says, or 6 methods like the test expects? Come on :)

File: lib/client/signalflow/request_manager.js [hunk 21]
```diff
@@ -390,11 +390,48 @@ function RequestManager(options) {
...
+      type: 'registerlivetail',
+      channel: requestId,
+      query: params.query,
+      throttleOptions: params.throttleOptions
+    }));
+  }
+
+  function stopLiveTail(requestId) {
+    if (knownComputations[requestId]) {
```
[the remark this answers is not in the document]
Comment (s2i-sfx) [reply lib/client/signalflow/request_manager.js:418]: It is doing this 
`      knownComputations[requestId].stopRequested = true;
`

Not sure if I need that. I am not setting stopRequested when I add my computation, so this might error

File: test/signalflow/request_manager.js (hunk 20 above)
Comment (s2i-sfx) [reply test/signalflow/request_manager.js:10]: fixed :)

File: lib/client/signalflow/signalflow_client.js (hunk 17 above)
Comment (s2i-sfx) [reply lib/client/signalflow/signalflow_client.js:112]: removed it

File: lib/client/signalflow/request_manager.js [hunk 22]
```diff
@@ -390,11 +390,48 @@ function RequestManager(options) {
     //destroy all open SSE connections
   }
 
+
+  function registerLiveTail(params, callbacks, overrideRequestId){
```
[the remark this answers is not in the document]
Comment (s2i-sfx) [reply lib/client/signalflow/request_manager.js:394]: I would like to keep the `onError` here because the BE plans to implement an error type message, so I can add it to request manager then

File: lib/client/signalflow/signalflow_client.js (hunk 18 above)
Comment (s2i-sfx) [reply lib/client/signalflow/signalflow_client.js:103]: I want to add this support in the future once the backend implements errors, so I left it here

File: lib/client/signalflow/signalflow_client.js (hunk 19 above)
Comment (s2i-sfx) [reply lib/client/signalflow/signalflow_client.js:77]: done

File: lib/client/signalflow/request_manager.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (mpetazzoni) [reply lib/client/signalflow/request_manager.js:419]: It won't, and it'll be harmless.

File: lib/client/signalflow/signalflow_client.js [hunk 23]
```diff
@@ -73,7 +73,67 @@ function SignalflowClient(apiToken, options) {
...
+ *                  matcher: {
+ *                    params: {
+ *                      op: string,
+ *                      args: object
+ *                    },
+ *                  }
+ *                },
+ *                throttleOptions: { rate: number }
+ *              }
+*/
```
Comment (mpetazzoni) [inline lib/client/signalflow/signalflow_client.js:90]: The indentation of that comment doesn't seem to match that of the function

File: lib/client/signalflow/signalflow_client.js (hunk 18 above)
Comment (mpetazzoni) [reply lib/client/signalflow/signalflow_client.js:117]: You misunderstood my point: you're taking care of that already at this layer in `resolveMessage()`.

File: lib/client/signalflow/request_manager.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (s2i-sfx) [reply lib/client/signalflow/request_manager.js:419]: done

File: lib/client/signalflow/signalflow_client.js (hunk 23 above)
Comment (s2i-sfx) [reply lib/client/signalflow/signalflow_client.js:90]: fixed

File: lib/client/signalflow/signalflow_client.js (hunk 18 above)
Comment (s2i-sfx) [reply lib/client/signalflow/signalflow_client.js]: I think I got what you are saying. I cleanup up the callbacks, please take a look at my latest commit
